tensorflow/python/ops/functional_ops.py

```diff
@@ -646,6 +646,207 @@ def compute(i, a_flat, tas):
+    name: (optional) Name prefix for the returned tensors.
+
+  Returns:
+    A tensor or (possibly nested) sequence of tensors.  Each tensor packs the
+    results of applying `fn` to tensors unpacked from `elems` along the first
+    dimension, and the previous accumulator value(s), from first to last.
+
+  Raises:
+    TypeError: if `fn` is not callable or the structure of the output of
+      `fn` and `initializer` do not match.
+    ValueError: if the lengths of the output of `fn` and `initializer`
+      do not match.
+
+  Examples:
+    ```python
+    elems = np.array([1, 2, 3, 4, 5, 6])
+    sum = scanr(lambda a, x: a + x, elems)
+    # sum == [720, 720, 360, 120, 30, 6]
+    ```
+
+    ```python
+    elems = np.array([1, 2, 3, 4, 5, 6])
+    initializer = np.array(0)
+    sum_one = scan(
+        lambda a, x: x[0] - x[1] + a, (elems + 1, elems), initializer)
+    # sum_one == [6, 5, 4, 3, 2, 1]
+    ```
+  """
+  if not callable(fn):
+    raise TypeError("fn must be callable.")
+
+  input_is_sequence = nest.is_sequence(elems)
+  input_flatten = lambda x: nest.flatten(x) if input_is_sequence else [x]
+  def input_pack(x):
+    return nest.pack_sequence_as(elems, x) if input_is_sequence else x[0]
+
+  if initializer is None:
+    output_is_sequence = input_is_sequence
+    output_flatten = input_flatten
+    output_pack = input_pack
+  else:
+    output_is_sequence = nest.is_sequence(initializer)
+    output_flatten = lambda x: nest.flatten(x) if output_is_sequence else [x]
+    def output_pack(x):
+      return (nest.pack_sequence_as(initializer, x)
+              if output_is_sequence else x[0])
+
+  elems_flat = input_flatten(elems)
+
+  in_graph_mode = not context.executing_eagerly()
+  with ops.name_scope(name, "scan", elems_flat):
+    # TODO(akshayka): Remove the in_graph_mode check once caching devices are
+    # supported in Eager
+    if in_graph_mode:
+      # Any get_variable calls in fn will cache the first call locally
+      # and not issue repeated network I/O requests for each iteration.
+      varscope = vs.get_variable_scope()
+      varscope_caching_device_was_none = False
+      if varscope.caching_device is None:
+        # TODO(ebrevdo): Change to using colocate_with here and in other
+        # methods.
+        varscope.set_caching_device(lambda op: op.device)
+        varscope_caching_device_was_none = True
+
+    # Convert elems to tensor array.
+    elems_flat = [
+        ops.convert_to_tensor(elem, name="elem") for elem in elems_flat]
+
+    # Convert elems to tensor array. n may be known statically.
+    n = elems_flat[0].shape[0].value or array_ops.shape(elems_flat[0])[0]
+
+    # TensorArrays are always flat
+    elems_ta = [
+        tensor_array_ops.TensorArray(dtype=elem.dtype, size=n,
+                                     dynamic_size=False,
+                                     infer_shape=True)
+        for elem in elems_flat]
+    # Unpack elements
```

Comment:
Here and everywhere: punctuate comments.

Reply:
@drpngx Thanks. Done.